pallets/pools/src/tranche.rs

```diff
@@ -0,0 +1,386 @@
+pub struct Tranche<Balance, Rate, Weight> {
+	pub(super) interest_per_sec: Rate,
+	pub(super) min_risk_buffer: Perquintill,
+	pub(super) seniority: Seniority,
+
+	pub(super) outstanding_invest_orders: Balance,
+	pub(super) outstanding_redeem_orders: Balance,
+
+	pub(super) debt: Balance,
+	pub(super) reserve: Balance,
+	pub(super) ratio: Perquintill,
+	pub(super) last_updated_interest: Moment,
+
+	pub(super) _phantom: PhantomData<Weight>,
+}
+
+impl<Balance, Rate, Weight> Tranches<Balance, Rate, Weight>
+where
+	Balance: Zero + Copy + BaseArithmetic,
+	Weight: Copy + From<u128>,
+	Rate: One,
+{
+	pub fn from_input(tranches: impl Iterator<Item = TrancheInput<Rate>>, now: Moment) -> Self {
+		Self(
+			tranches
+				.enumerate()
+				.map(|(id, input)| Tranche {
+					interest_per_sec: input.interest_per_sec.unwrap_or(One::one()),
+					min_risk_buffer: input.min_risk_buffer.unwrap_or(Perquintill::zero()),
+					// seniority increases as index since the order is from junior to senior
+					seniority: input
+						.seniority
+						.unwrap_or(id.try_into().expect("MaxTranches is u32")),
+					outstanding_invest_orders: Zero::zero(),
+					outstanding_redeem_orders: Zero::zero(),
+
+					debt: Zero::zero(),
+					reserve: Zero::zero(),
+					ratio: Perquintill::zero(),
+					last_updated_interest: now,
+					_phantom: Default::default(),
+				})
+				.collect(),
+		)
+	}
+
+	pub fn num_tranches(&self) -> usize {
+		self.0.len()
+	}
+
+	pub fn into_tranches(self) -> Vec<Tranche<Balance, Rate, Weight>> {
+		self.0
+	}
+
+	pub fn as_tranche_slice(&self) -> &[Tranche<Balance, Rate, Weight>] {
+		self.0.as_slice()
+	}
+
+	pub fn as_mut_tranche_slice(&mut self) -> &mut [Tranche<Balance, Rate, Weight>] {
+		self.0.as_mut_slice()
+	}
+
+	pub fn supplies(&self) -> Result<Vec<Balance>, DispatchError> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.debt.checked_add(&tranche.reserve))
+			.collect::<Option<Vec<_>>>()
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn acc_supply(&self) -> Result<Balance, DispatchError> {
+		self.0
+			.iter()
+			.fold(Some(Balance::zero()), |sum, tranche| {
+				sum.and_then(|acc| {
+					acc.checked_add(&tranche.debt)
+						.and_then(|acc| acc.checked_add(&tranche.reserve))
+				})
+			})
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn investments(&self) -> Vec<Balance> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.outstanding_invest_orders)
+			.collect()
+	}
+
+	pub fn acc_investments(&self) -> Result<Balance, DispatchError> {
+		self.0
+			.iter()
+			.fold(Some(Balance::zero()), |sum, tranche| {
+				sum.and_then(|acc| acc.checked_add(&tranche.outstanding_invest_orders))
+			})
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn redemptions(&self) -> Vec<Balance> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.outstanding_redeem_orders)
+			.collect()
+	}
+
+	pub fn acc_redemptions(&self) -> Result<Balance, DispatchError> {
```

Comment:
Can we rename this to `total_outstanding_redeem_orders`?

Reply:
What is the difference between the `outstanding_invest` of the normal `Tranche` and the `invest` of the `EpochExecutionTranche` fields? From a business logic perspective.